File: app/src/shared/lib/crypto.js
// Single-key, application-level field encryption (AES-256-GCM via Web Crypto).
//
// One app-wide key encrypts every user's content fields so that a DB leak yields
// only ciphertext. The key never lives in the database or the JS bundle — it is
// served to signed-in users by an authenticated edge function (see lib/encKey.js).
//
// Blob format:  "v1:" + base64(iv) + ":" + base64(ciphertext)
// Values without the "v1:" prefix are treated as plaintext. That keeps reads safe
// for (a) rows written before encryption was turned on and (b) deliberately-public
// shared pages, which we store plaintext on purpose.

const PREFIX = 'v1:';
const IV_BYTES = 12; // 96-bit nonce, the standard size for AES-GCM

function b64encode(bytes) {
  const arr = new Uint8Array(bytes);
  let s = '';
  for (let i = 0; i < arr.length; i++) s += String.fromCharCode(arr[i]);
  return btoa(s);
}

function b64decode(str) {
  const bin = atob(str);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

// Import a base64-encoded 32-byte key into a non-extractable AES-GCM CryptoKey.
export async function importKey(base64Key) {
  const raw = b64decode(base64Key);
  if (raw.length !== 32) {
    throw new Error('Encryption key must be 32 bytes once base64-decoded.');
  }
  return crypto.subtle.importKey(
    'raw', raw, { name: 'AES-GCM' }, false, ['encrypt', 'decrypt'],
  );
}

export function isEncrypted(value) {
  return typeof value === 'string' && value.startsWith(PREFIX);
}

// A fresh random IV on every save would make every field look "changed" to the
// sync diff, re-uploading everything on each keystroke. Caching the blob by its
// plaintext keeps unchanged fields byte-identical across pushes, so the diff only
// ships fields the user actually edited. Same plaintext always maps to the same
// IV here, which is safe because an IV is only ever reused with its own plaintext.
const blobByPlain = new Map();

export async function encryptText(key, text) {
  const s = text == null ? '' : String(text);
  const cached = blobByPlain.get(s);
  if (cached) return cached;
  const iv = crypto.getRandomValues(new Uint8Array(IV_BYTES));
  const ct = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv }, key, new TextEncoder().encode(s),
  );
  const blob = PREFIX + b64encode(iv) + ':' + b64encode(ct);
  blobByPlain.set(s, blob);
  return blob;
}

export async function decryptText(key, blob) {
  if (!isEncrypted(blob)) return blob == null ? '' : blob; // plaintext passthrough
  const [, ivB64, ctB64] = blob.split(':');
  const pt = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: b64decode(ivB64) }, key, b64decode(ctB64),
  );
  const s = new TextDecoder().decode(pt);
  blobByPlain.set(s, blob); // seed the cache so re-pushing this value won't churn
  return s;
}

// ----- Array / structured helpers -----

// tags is a `text[]` column, so encrypt each element to keep it a real array.
export async function encryptTags(key, tags, plaintext) {
  const arr = Array.isArray(tags) ? tags : [];
  if (plaintext) return arr;
  return Promise.all(arr.map(t => encryptText(key, t)));
}

export async function decryptTags(key, tags) {
  const arr = Array.isArray(tags) ? tags : [];
  return Promise.all(arr.map(t => decryptText(key, t)));
}

// subtasks is a jsonb column of objects; encrypt the whole array into one blob
// and store it as a JSON string scalar.
export async function encryptJson(key, value, plaintext) {
  const arr = Array.isArray(value) ? value : [];
  if (plaintext) return arr;
  return encryptText(key, JSON.stringify(arr));
}

export async function decryptJson(key, value) {
  if (isEncrypted(value)) {
    try { return JSON.parse(await decryptText(key, value)); }
    catch { return []; }
  }
  return Array.isArray(value) ? value : []; // plaintext passthrough
}

File: mobile/notemd/notemd/Crypto.swift
//
//  Crypto.swift
//  notemd
//
//  Application-level field encryption, byte-compatible with the web client
//  (app/src/lib/crypto.js). AES-256-GCM via CryptoKit.
//
//  Blob format:  "v1:" + base64(iv) + ":" + base64(ciphertext+tag)
//  Values without the "v1:" prefix are plaintext and pass through untouched
//  (covers pre-encryption rows and deliberately-public pages).
//
//  GCM tag gotcha: Web Crypto appends the 16-byte auth tag to the ciphertext,
//  while CryptoKit keeps ciphertext and tag separate. So on decrypt we split the
//  last 16 bytes off as the tag; on encrypt we concatenate ciphertext + tag.
//

import Foundation
import CryptoKit

enum NoteMDCrypto {
    private static let prefix = "v1:"
    private static let ivBytes = 12      // 96-bit nonce, standard for AES-GCM
    private static let tagBytes = 16

    static func isEncrypted(_ value: String?) -> Bool {
        value?.hasPrefix(prefix) ?? false
    }

    // MARK: - Text

    static func encryptText(_ key: SymmetricKey, _ text: String?) throws -> String {
        let s = text ?? ""
        let nonce = AES.GCM.Nonce()  // 12 random bytes
        let box = try AES.GCM.seal(Data(s.utf8), using: key, nonce: nonce)
        let ctTag = box.ciphertext + box.tag
        let ivB64 = Data(nonce).base64EncodedString()
        return prefix + ivB64 + ":" + ctTag.base64EncodedString()
    }

    static func decryptText(_ key: SymmetricKey, _ blob: String?) -> String {
        guard let blob, blob.hasPrefix(prefix) else { return blob ?? "" }
        // split into ["v1", ivB64, ctB64] — ct may itself contain no extra ":"
        let body = String(blob.dropFirst(prefix.count))
        guard let sep = body.firstIndex(of: ":") else { return "" }
        let ivB64 = String(body[body.startIndex..<sep])
        let ctB64 = String(body[body.index(after: sep)...])
        guard
            let iv = Data(base64Encoded: ivB64),
            let ctAndTag = Data(base64Encoded: ctB64),
            ctAndTag.count >= tagBytes
        else { return "" }

        let tag = ctAndTag.suffix(tagBytes)
        let ct = ctAndTag.prefix(ctAndTag.count - tagBytes)
        do {
            let box = try AES.GCM.SealedBox(nonce: try AES.GCM.Nonce(data: iv),
                                            ciphertext: ct, tag: tag)
            let pt = try AES.GCM.open(box, using: key)
            return String(decoding: pt, as: UTF8.self)
        } catch {
            return ""
        }
    }

    // MARK: - Structured (subtasks jsonb scalar)

    // Matches encryptJson/decryptJson: JSON-encode the array, then encrypt to a
    // single scalar string stored in the jsonb column.
    static func encryptJSON<T: Encodable>(_ key: SymmetricKey, _ value: T) throws -> String {
        let data = try JSONEncoder().encode(value)
        let json = String(decoding: data, as: UTF8.self)
        return try encryptText(key, json)
    }

    static func decryptJSON<T: Decodable>(_ key: SymmetricKey, _ value: Any?, as type: T.Type) -> T? {
        // Encrypted scalar string → decrypt then decode.
        if let s = value as? String, isEncrypted(s) {
            let json = decryptText(key, s)
            return try? JSONDecoder().decode(T.self, from: Data(json.utf8))
        }
        // Plaintext passthrough: already-parsed array (legacy rows).
        if let raw = value {
            if let data = try? JSONSerialization.data(withJSONObject: raw) {
                return try? JSONDecoder().decode(T.self, from: data)
            }
        }
        return nil
    }

    /// Import a base64-encoded 32-byte key (from the enc-key function) as a key.
    static func importKey(base64 key: String) throws -> SymmetricKey {
        guard let raw = Data(base64Encoded: key), raw.count == 32 else {
            throw NoteMDError.message("Encryption key must be 32 bytes once base64-decoded.")
        }
        return SymmetricKey(data: raw)
    }
}

enum NoteMDError: LocalizedError {
    case message(String)
    var errorDescription: String? {
        if case let .message(m) = self { return m }
        return nil
    }
}
